### DISCUSSION TITLE
Request for Feedback: Introducing a CLA or DCO to MapLibre Projects?

### DISCUSSION THREAD

**Discussion by louwers** [maplibre/maplibre#274] [General]
MapLibre GL JS and MapLibre Native are both forks of now proprietary non-free software developed by a large corporation. We need to be especially careful that code contributions do not violate any copyright.

We are somewhat covered by the [GitHub Terms of Service](https://docs.github.com/en/site-policy/github-terms/github-terms-of-service), which states:

> Whenever you add Content to a repository containing notice of a license, you license that Content under the same terms, and you agree that you have the right to license that Content under those terms. If you have a separate agreement to license that Content under different terms, such as a contributor license agreement, that agreement will supersede.
> 
> Isn't this just how it works already? Yep. This is widely accepted as the norm in the open-source community; it's commonly referred to by the shorthand "inbound=outbound". We're just making it explicit.

We can protect the project a bit more by either introducing a Contributor License Agreement (CLA) or a Developer Certificate of Origin (DCO). Before contributing you explicitly state that you have the rights to license your contribution. In the case of a CLA, you do it once, in the case of a DCO, you need to sign every commit.

Some resources:

- [CLA vs. DCO: What's the difference?](https://opensource.com/article/18/3/cla-vs-dco-whats-difference)
- [Developer Certificate of Origin](https://developercertificate.org/)
- [Apache Contributor License Agreement Text](https://www.apache.org/licenses/contributor-agreements.html): this is what a lot of Contributor License Agreements are based on.
- DCO, used by e.g. Nextcloud, Linux, OpenSearch 
- CLA, used by e.g. Qt, Apache projects, Python, Meta and Google projects

We discussed this topic during today's technical steering committee meetings. This discussion thread is intended to gather feedback from members in the MapLibre community about this topic. If you work at a company using MapLibre, it would also be helpful to share how your legal department thinks about this topic.

After this discussion the governing board can (re-)consider if we should introduce either of the two mechanisms or keep the status quo.

**Comment by louwers** [comment]
Personally I don't think adding either a CLA or DCO is worth the added barrier to contributing. Although we do need to remain very vigilant on this topic, especially in the case of MapLibre GL JS where the project it forked from is now source available but proprietary.

@klokan  also mentioned that the advising lawyers when MapLibre was founded as a project did not explicitly recommend introducing a CLA.

**Reply by HarelM** [reply]
You mean @klokan 😃

**Reply by louwers** [reply]
Oops wrong Petr. Yes. 😉

**Reply by ianthetechie** [reply]
I'm having a bit of trouble parsing the double negative in your first sentence ;) I think that was accidental and you mean to say that you do NOT think it's a necessary step at this time as it's an extra barrier to contributing.

As the GitHub terms state, this is already the way things work, and I agree that a CLA is quite a burden. I just looked over Apache's and it seems like it would be a rather large hassle to manage as we would need to collect personal info and then have a privacy policy around that etc. as it's a formal contract. The DCO on the other hand reads a lot more like a BSD license and is plain-english and easy to understand. If we were to go with either, I'd suggest the DCO.

**Reply by louwers** [reply]
@ianthetechie Fixed and thanks for your input!

**Comment by pnorman** [comment]
> Whenever you add Content to a repository containing notice of a license, you license that Content under the same terms, and you agree that you have the right to license that Content under those terms.

It's worth noting that someone publishing a branch and doing a pull request is explicitly publishing their code under the license of the repo by the act of publishing the branch. What else is needed? Unless we're taking patches as contributions, every contribution includes a license.

A CLA would exclude some contributors because they are difficult to get signed in some cases, and others are reluctant to contribute to a project which requires assigning rights beyond the project license.

**Reply by louwers** [reply]
Thanks for your input. 👍 

As I understand it, an additional right you grant with a CLA is the right to re-license. This is not relevant in our case, because MapLibre GL JS and MapLibre Native are not going to be (and cannot be) relicensed anyway.

And yes, fully agree that there is a risk that some people may not want or may not be able to sign a CLA, which would be bad for the project.
